vsintegration/src/FSharp.Editor/Options/SettingsPersistence.fs

```diff
@@ -18,42 +18,67 @@ type SVsSettingsPersistenceManager = class end
+          typ.Namespace + "." + typ.Name ]
+        
+    let storageKey (typ: Type) = storageKeyVersions typ |> List.head
 
     // Each group of settings is a value of some named type, for example 'IntelliSenseOptions', 'QuickInfoOptions'
-    // We cache exactly one instance of each, treating them as immutable.
-    // This cache is updated by the SettingsStore when the user changes an option.
-    let cache = System.Collections.Concurrent.ConcurrentDictionary<Type, obj>()
+    // and it is usually representing one separate option page in the UI.
+    // We cache exactly one immutable value of each type.
+    // This cache is updated by the SettingsStore when the user makes changes in the Options dialog
+    // or when a change is propagated from another VS IDE instance by SVsSettingsPersistenceManager. 
+    let cache = ConcurrentDictionary<Type, obj>()
 
-    let read() =
+    let getCached() =
         match cache.TryGetValue(typeof<'t>) with
-        | true, value -> value :?> 't
-        | _ -> failwithf "Settings %s are not registered." typeof<'t>.Name 
+        | true, (:? 't as value) -> value
+        | _ -> failwithf "Settings %s are not registered." typeof<'t>.Name
 
-    let write settings = cache.[settings.GetType()] <- settings
+    let keepInCache settings = cache.[settings.GetType()] <- settings
+
+    // The settings record, even though immutable, is being effectively mutated in two instances:
+    //   when it is passed to the UI (provided it is marked with CLIMutable attribute);
+    //   when it is being populated from JSON using JsonConvert.PopulateObject;
+    // We make a deep copy in these instances to isolate and contain the mutation
+    let clone (v: 't) = JsonConvert.SerializeObject v |> JsonConvert.DeserializeObject<'t>
 
     let updateFromStore settings =
-        let result, json = settings.GetType() |> storageKey |> settingsManager.TryGetValue
-        if result = GetValueResult.Success then
-            // if it fails we just return what we got
-            try JsonConvert.PopulateObject(json, settings) with _ -> () 
-        settings
-        
-    member __.Read() = read()
+        // make a deep copy so that PopulateObject does not alter the original
+        let copy = clone settings
+        // if the new key is not found by ISettingsManager, we try the old keys
+        // so that user settings are not lost
+        settings.GetType() |> storageKeyVersions 
+        |> Seq.map (settingsManager.TryGetValue)
+        |> Seq.tryPick ( function GetValueResult.Success, json -> Some json | _ -> None )
+        |> Option.iter (fun json -> try JsonConvert.PopulateObject(json, copy) with _ -> ())
+        copy
+
+    member __.Get() = getCached()
+
+    // Used by the AbstractOptionPage to populate dialog controls.
+    // We clone the value because it may be altered by the UI if declared with [<CLIMutable>]
+    member __.Read() = getCached() |> clone
```

Comment:
Having `Get` and `Read` feels confusing. We should really only have `Read` like we did before.

Reply:
When calling `Read`, is there a reason we can't just read directly from the `settingsManager` and deserialize everytime? If performance is the reason, then it's ok that we cache, but when does the cache get invalidated? 

After looking at `ISettingsManager`, I feel like when we call `Read` for the first time, we should call `settingsManager.GetSubset(*storage key*)` and subscribe to the setting changes event so we can invalidate our cache; then we grab the most recent value calling `settingsManager.GetValueOrDefault`, then cache that value. That way whenever we call `Read`, we just grab it from the cache. When the settings changed event fires, we invalidate our cache based on storage key, so on the next `Read` we get the new value and cache it again. What are your thoughts?

Reply:
Yes, this method is now definitely misnamed. It should be named GetClone or something, as it's there just for the sake of the UI implementation that we have.
The overall idea is that we subscribe to changes in the Register method and it is the only place we get the settings proactively from settingsManager (that happens one time) and then let the event handler update the cache, so we have everything always in sync, that's basically what you propose, I think.

Reply:
Ah, I didn't realize we already listen for a settings changed event when we register. That makes sense. In that case, we should just invalidate the cache in that handler.